### PR TITLE
svc: Fully implement svcSignalToAddress and svcWaitForAddress

### PR DESCRIPTION
**Pull request by SciresM**
These SVCs were introduced in 4.0.0, and are used by newer games to implement intra-process Event primitives (Mario Tennis Aces uses them, for example).

I also modified the "MisalignedAddress" error code to reflect its true meaning as "InvalidAddress", and further filled in several ERR_ definitions to return console-accurate codes.

### CODE REVIEW

File: src/core/hle/kernel/address_arbiter.cpp [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+ResultCode WaitForAddress(VAddr address, s64 timeout) {
+    SharedPtr<Thread> current_thread = GetCurrentThread();
+    current_thread->arb_wait_address = address;
+    current_thread->status = THREADSTATUS_WAIT_ARB;
+    current_thread->wakeup_callback = nullptr;
+
+    current_thread->WakeAfterDelay(timeout);
+
+    Core::System::GetInstance().CpuCore(current_thread->processor_id).PrepareReschedule();
+    // This should never actually execute.
```
Comment (Subv) [inline src/core/hle/kernel/address_arbiter.cpp]: What's with this comment?

File: src/core/hle/kernel/address_arbiter.cpp [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+// Wake up num_to_wake (or all) threads in a vector.
+void WakeThreads(std::vector<SharedPtr<Thread>>& waiting_threads, s32 num_to_wake) {
+    // Only process up to 'target' threads, unless 'target' is <= 0, in which case process
+    // them all.
+    size_t last = waiting_threads.size();
+    if (num_to_wake > 0)
+        last = num_to_wake;
+
+    // Signal the waiting threads.
+    // TODO: Rescheduling should not occur while waking threads. How can it be prevented?
```
Comment (Subv) [inline src/core/hle/kernel/address_arbiter.cpp]: You can remove this comment

File: src/core/hle/kernel/address_arbiter.cpp [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+// Signals an address being waited on and increments its value if equal to the value argument.
+ResultCode IncrementAndSignalToAddressIfEqual(VAddr address, s32 value, s32 num_to_wake) {
+    // Ensure that we can write to the address.
+    if (!Memory::IsValidVirtualAddress(address)) {
+        return ERR_INVALID_ADDRESS_STATE;
+    }
+
+    if ((s32)Memory::Read32(address) == value) {
+        Memory::Write32(address, (u32)(value + 1));
```
Comment (Subv) [inline src/core/hle/kernel/address_arbiter.cpp]: static_cast

File: src/core/hle/kernel/address_arbiter.cpp [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+    if (waiting_threads.size() == 0) {
+        updated_value = value - 1;
+    } else if (num_to_wake <= 0 || waiting_threads.size() <= num_to_wake) {
+        updated_value = value + 1;
+    } else {
+        updated_value = value;
+    }
+
+    if ((s32)Memory::Read32(address) == value) {
+        Memory::Write32(address, (u32)(updated_value));
```
Comment (Subv) [inline src/core/hle/kernel/address_arbiter.cpp]: static_cast

File: src/core/hle/kernel/address_arbiter.cpp [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+// Waits on an address if the value passed is less than the argument value, optionally decrementing.
+ResultCode WaitForAddressIfLessThan(VAddr address, s32 value, s64 timeout, bool should_decrement) {
+    // Ensure that we can read the address.
+    if (!Memory::IsValidVirtualAddress(address)) {
+        return ERR_INVALID_ADDRESS_STATE;
+    }
+
+    s32 cur_value = (s32)Memory::Read32(address);
+    if (cur_value < value) {
+        Memory::Write32(address, (u32)(cur_value - 1));
```
Comment (Subv) [inline src/core/hle/kernel/address_arbiter.cpp]: static_cast

File: src/core/hle/kernel/svc.cpp [hunk 6]
```diff
@@ -689,6 +690,58 @@ static ResultCode SignalProcessWideKey(VAddr condition_variable_addr, s32 target
     return RESULT_SUCCESS;
 }
 
+// Wait for an address (via Address Arbiter)
+static ResultCode WaitForAddress(VAddr address, u32 type, s32 value, s64 timeout) {
+    NGLOG_WARNING(Kernel_SVC, "called, address=0x{:X}, type=0x{:X}, value=0x{:X}, timeout={}",
+                  address, type, value, timeout);
+    // If the passed address is a kernel virtual address, return invalid memory state.
+    if ((address + 0x8000000000LL) < 0x7FFFE00000LL) {
```
Comment (Subv) [inline src/core/hle/kernel/svc.cpp]: Please add a function for this, and don't use magic numbers

File: src/core/hle/kernel/svc.cpp [hunk 7]
```diff
@@ -689,6 +690,58 @@ static ResultCode SignalProcessWideKey(VAddr condition_variable_addr, s32 target
...
+        return ERR_INVALID_ADDRESS_STATE;
+    }
+    // If the address is not properly aligned to 4 bytes, return invalid address.
+    if (address % sizeof(u32) != 0) {
+        return ERR_INVALID_ADDRESS;
+    }
+
+    switch ((AddressArbiter::SignalType)type) {
```
Comment (Subv) [inline src/core/hle/kernel/svc.cpp]: static_cast

File: src/core/hle/kernel/svc.cpp [hunk 8]
```diff
@@ -689,6 +690,58 @@ static ResultCode SignalProcessWideKey(VAddr condition_variable_addr, s32 target
...
+    // If the passed address is a kernel virtual address, return invalid memory state.
+    if ((address + 0x8000000000LL) < 0x7FFFE00000LL) {
+        return ERR_INVALID_ADDRESS_STATE;
+    }
+    // If the address is not properly aligned to 4 bytes, return invalid address.
+    if (address % sizeof(u32) != 0) {
+        return ERR_INVALID_ADDRESS;
+    }
+
+    switch ((AddressArbiter::ArbitrationType)type) {
```
Comment (Subv) [inline src/core/hle/kernel/svc.cpp]: static_cast

File: src/core/hle/kernel/address_arbiter.cpp [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+
+    current_thread->WakeAfterDelay(timeout);
+
+    Core::System::GetInstance().CpuCore(current_thread->processor_id).PrepareReschedule();
+    // This should never actually execute.
+    return RESULT_SUCCESS;
+}
+
+// Gets the threads waiting on an address.
+void GetThreadsWaitingOnAddress(std::vector<SharedPtr<Thread>>& waiting_threads, VAddr address) {
```
Comment (Subv) [inline src/core/hle/kernel/address_arbiter.cpp]: Those functions which are not part of the public interface should be static

File: src/core/hle/kernel/address_arbiter.cpp [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+    } else {
+        return ERR_INVALID_STATE;
+    }
+
+    WakeThreads(waiting_threads, num_to_wake);
+    return RESULT_SUCCESS;
+}
+
+// Waits on an address if the value passed is less than the argument value, optionally decrementing.
+ResultCode WaitForAddressIfLessThan(VAddr address, s32 value, s64 timeout, bool should_decrement) {
```
Comment (lioncash) [inline src/core/hle/kernel/address_arbiter.cpp]: I'm a little late to the review, but `should_decrement` isn't used within the function, is this intentional?

File: src/core/hle/kernel/address_arbiter.cpp [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+
+    // Get threads waiting on the address.
+    std::vector<SharedPtr<Thread>> waiting_threads;
+    GetThreadsWaitingOnAddress(waiting_threads, address);
+
+    // Determine the modified value depending on the waiting count.
+    s32 updated_value;
+    if (waiting_threads.size() == 0) {
+        updated_value = value - 1;
+    } else if (num_to_wake <= 0 || waiting_threads.size() <= num_to_wake) {
```
Comment (lioncash) [inline src/core/hle/kernel/address_arbiter.cpp]: `waiting_threads.size() <= num_to_wake` is an `unsigned <= signed` comparison which can cause warnings.

**Pull request merged**